trace-agent: more permissive trace normalization with metrics & logs

djova opened the pull request:
### What does this PR do?

The trace agent now lets through a wider variety of traces, automatically correcting some malformed traces instead of dropping them. The quantity of malformed traces is kept track of in a new `datadog.trace_agent.receiver.spans_malformed` metric (tagged by `reason`). 

* The following fields are now replaced with reasonable defaults if invalid or empty and truncated if exceeding max length: `span.service`, `span.name`, `span.resource`, `span.type`.
* `span.duration=0` is now allowed.
* Missing span start date now defaults to current time.

Traces are now dropped only for the following reasons: `empty_trace`, `foreign_span`, with the `datadog.trace_agent.receiver.traces_dropped` metric tagged accordingly.

### Motivation

Goal is to redefine the Agent ingestion so that we ingest as many traces as we can, surfacing malformed traces in the UI instead of dropping them at the agent.

djova committed:
trace-agent-normalization

djova committed:
more

djova committed:
rename

djova committed:
.

djova committed:
.

djova committed:
.

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -119,8 +160,9 @@ func normalize(s *pb.Span) error {
 	if sc, ok := s.Meta["http.status_code"]; ok {
 		if !isValidStatusCode(sc) {
+			atomic.AddInt64(&ts.TracesMalformedInvalidHttpStatusCode, 1)
+			log.Debugf("fixing malformed trace (reason:invalid_http_status_code), dropping invalid http.status_code=%s: %s", sc, s)
```
Please capitalize all log messages.

djova replied:
Done

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -244,6 +339,7 @@ func (s *Stats) isEmpty() bool {
 
 // String returns a string representation of the Stats struct
 func (s *Stats) String() string {
+	// TODO: just marshal this whole thing to JSON
```
Why? This is used for outputting stats.

djova replied:
Just a random thought while writing, not relevant anymore, removed

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +189,66 @@ func (ts *TagStats) publish() {
 type Stats struct {
 	// TracesReceived is the total number of traces received, including the dropped ones.
 	TracesReceived int64
+	// TracesDroppedDecodingError is the number of traces dropped due to decoding error
```
This is really nasty. Would it be better if we used embedded structs? Perhaps:
```go
type TracesDropped struct {
    DecodingError int64
    EmptyTrace int64
    TraceIDZero
    // ...
}
```
It might be a bit better.

The same for `TracesMalformed`.

djova replied:
Agreed, updated

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +189,66 @@ func (ts *TagStats) publish() {
+
+
+	var droppedReasons []string
+	for _, c := range ts.droppedTraceReasonCounts() {
```
This whole thing is a bit too "smart" and slightly confusing. That structure is basically a `map[string]int64`, so let's make it super obvious to a reader what's going on here. This package is the definition of verbose already so let's do the same here:
```go
for reason, count := range map[string]int64{
	"reason:decoding_error": atomic.LoadInt64(&ts.TracesDroppedDecodingError),
	"reason:empty_trace":    atomic.LoadInt64(&ts.TracesDroppedEmptyTrace),
	"reason:trace_id_zero":  atomic.LoadInt64(&ts.TracesDroppedTraceIdZero),
	"reason:span_id_zero":   atomic.LoadInt64(&ts.TracesDroppedSpanIdZero),
	"reason:foreign_span":   atomic.LoadInt64(&ts.TracesDroppedForeignSpan),
} {
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", count, append(tags, reason), 1)
}
```
...then we don't have to scroll and jump aroind the file to see which struct field belongs to which metric.

djova replied:
Yeah it was confusing. I changed it since then, can you take another look?

gbbr replied:
I would prefer it if you used the above suggestion. It significantly reduces the amount of code. I don't think that the extra type and those functions bring any value.

djova replied:
Oh I see have another struct `ts.TracesDropped`, yeah that's a bit nicer. Will do

gbbr replied:
> Oh I see have another struct ts.TracesDropped, yeah that's a bit nicer. Will do
That's the discussion in [this thread](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r296568295) 😄

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +189,66 @@ func (ts *TagStats) publish() {
+	}
+
+	var malformedReasons []string
+	for _, c := range ts.malformedTraceReasonCounts() {
```
Same here.

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +189,66 @@ func (ts *TagStats) publish() {
+		}
+	}
+
+	var normalizerMessages []string
```
Can you please remove this entire logging logic. This is not the right place to do it. Let's let `publish` publish only stats. The receiver stats are published once per minute [here](https://github.com/DataDog/datadog-agent/blob/bd9c45ba2f055b9349cc968884c498776279d24e/pkg/trace/api/api.go#L395-L397) and they are the result of the `fmt.Stringer` implementation from [here](https://github.com/DataDog/datadog-agent/blob/bd9c45ba2f055b9349cc968884c498776279d24e/pkg/trace/info/stats.go#L246). All you have to do is improve the `String` method a bit to include dropped and malformed traces and it'll be 👍

djova replied:
Yeah I planned on moving it elsewhere anyway, just had it there while I was working on it

gbbr commented:
Left some pointers. Happy to review again once you feel it's ready.

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -32,37 +33,76 @@ var (
+		log.Debugf("dropping invalid trace (reason:span_id_zero): %s", s)
+		return errors.New("span_id_zero")
+	}
+	if s.TraceID != firstSpan.TraceID {
```
The signature of this function becomes confusing now. Why can't we leave this comparison where it was? In the `for` loop? It seems redundant to pass what is basically a trace ID to a function, as an entire span, only to compare the current span's trace ID to it.

djova replied:
Good point, on second thought this does seem confusing. Moved the foreign span check to outside `normalize` so we don't need to pass the firstSpan in anymore.

djova committed:
tests working

djova committed:
cleanup

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -2,23 +2,29 @@ package api
 )
 
-func testSpan() *pb.Span {
+func newTagStats() *info.TagStats {
```
```
var ts info.TagStats
```
Has the exact same outcome. You don't need this function. Then simply use `&ts` if you need a pointer form.

djova replied:
True, updated

djova committed:
break out logging

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +39,315 @@ func testSpan() *pb.Span {
+	}
+}
+
 func TestNormalizeOK(t *testing.T) {
```
If you can spare the time and don't mind too much. It would be fantastic to make one big test here called `TestNormalize` and all of the rest sub-tests of it. This is optional but would be a very much appreciated clean up.

As a side-effect, you could add the two helper functions from above inside the parent function, making it clear what their scope is, e.g.:
```go
func TestNormalize(t *testing.T) {
    assertNoIssues := func(.. ) { .... }
    assertIssue := func(...) { ... }
    
    t.Run("ok", ...
    ...
}
```

djova replied:
I just tried doing it and I don't think it ends up any cleaner. 

All the tests in `normalizer_test.go` are related to normalization, and each test case is a simple independent test case. So if we grouped them all into subtests we still end up with the same amount of code, and it maybe even gets a bit more complicated due having these test cases declared dynamically vs have them be statically declared as functions. 

Use subtests might make more sense if we could statically declare these test cases in some way, like [in the go examples](https://blog.golang.org/subtests). I don't think we can do that here though.

gbbr replied:
I understand. Ok, makes sense. Let's leave it then.

djova committed:
break out into separate structs

djova committed:
remove unnecessary type

djova committed:
json

djova committed:
capitalize error messages

djova committed:
info vs warn stats

djova committed:
trace/api: add missing expvar handler and fix silent test fail

Changes:
* restore missing `/debug/vars` endpoint and add corresponding test
* fix silent test failures in `pkg/trace/api/...`

`/debug/vars` was accidentally removed when we switched to a custom mux handler in https://github.com/DataDog/datadog-agent/pull/3308. We partially restored the lost endpoints in https://github.com/DataDog/datadog-agent/pull/3365. This restores another one.

Added an api test to confirm the existence of the expvar endpoint. It now fails if it's not there:

```
--- FAIL: TestExpvar (0.00s)
    api_test.go:764:
        	Error Trace:	api_test.go:764
        	Error:      	Not equal:
        	            	expected: 404
        	            	actual  : 200
        	Test:       	TestExpvar
        	Messages:   	failed to read expvars from local server
```

In order for the test to work `trace/api/api_test.go:TestMain` had to be fixed to call `os.Exit(m.Run())` instead of `m.Run()`. When calling only `m.Run()` it causes all tests in `pkg/trace/api/...` to fail silently. As per [the docs](https://golang.org/pkg/testing/#hdr-Main), *It should then call os.Exit with the result of m.Run*.

We're seeing this problem in the latest RC that we're running. When running `datadog-agent status` it shows an error with the expvar check:

```
      Instance ID: go_expvar:f77c6c92ba32d2ef [ERROR]
      Total Runs: 59,076
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 4ms
      Error: HTTPConnectionPool(host='localhost', port=7654): Max retries exceeded with url: /debug/vars (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fda1a0e9c50>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

djova committed:
.

djova committed:
fix test

djova committed:
simplify

djova committed:
remove newTagStats

djova committed:
sorted reason string output

djova committed:
linting and cleanup

djova commented:
Depends on https://github.com/DataDog/datadog-agent/pull/3747

djova committed:
warnStrings formatting

djova commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
+
+// statsStructToMap marshals the provided struct into a map[string]int64 by first marshaling it into json
+// then marshing from json into the map
+func statsStructToMap(v interface{}, name string) (result map[string]int64) {
```
@gbbr, what do you think of this? It's kind of hacky... but it is a nice way to have the source of truth for the normalization `reason` names be in one place, the json-annotations, and then we use that everywhere else without having to duplicate: reason tags for metrics, agent info page, periodic log message.

gbbr replied:
I agree. This _is_ quite hacky. The thing that you are trying to save here (one place for tags) is not worth the cost of the complexity and time it takes a reader to try and figure out this code. Furthermore, it also adds JSON struct fields to the type, creating even more confusion, leading people to believe that this is marshaled (which it is, but for the wrong reasons). 

Honestly, it also crossed my mind that it would be nice to have struct fields which represent the tags that these stats belong to. And this is doable, with custom struct fields, but I'm not sure if it's worth the effort. And even if we were to do it, then we should be doing it for all fields and simply include the metric and its tags.

If you like, I can suggest two ways to do this which are more "Go-like". 

1. Keep one publish method on each of these structs, which publishes the metrics just like the already existing one. For example:
```go
// publish publishes and resets the accumulated metrics.
func (td *TracesDropped) publish(tags []string) {
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.DecodingError), append(tags, "reason:decoding_error"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.EmptyTrace), append(tags, "reason:empty_trace"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.TraceIDZero), append(tags, "reason:trace_id_zero"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.SpanIDZero), append(tags, "reason:span_id_zero"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.ForeignSpan), append(tags, "reason:foreign_span"), 1)
}
```
This makes it clear to everyone because it behaves just like `(*TagStats).publish`. It would then be called there and receive the additional tags.

2. Use the suggestion given [here](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r296569285). Personally, I prefer this one because it is very linear and easy to follow.

Lastly, I think a good practice in Go would be to have all these tags as constants, for example:
```go
const (
	tagReasonDuplicateSpanID       = "reason:duplicate_span_id"
	tagReasonServiceEmpty          = "reason:service_empty"
	tagReasonServiceTruncate       = "reason:service_truncate"
	tagReasonServiceInvalid        = "reason:service_invalid"
	tagReasonSpanNameEmpty         = "reason:span_name_empty"
	tagReasonSpanNameTruncate      = "reason:span_name_truncate"
	tagReasonSpanNameInvalid       = "reason:span_name_invalid"
	tagReasonResourceEmpty         = "reason:resource_empty"
	tagReasonTypeTruncate          = "reason:type_truncate"
	tagReasonInvalidStartDate      = "reason:invalid_start_date"
	tagReasonInvalidDuration       = "reason:invalid_duration"
	tagReasonInvalidHTTPStatusCode = "reason:invalid_http_status_code"
	tagReasonDecodingError         = "reason:decoding_error"
	tagReasonEmptyTrace            = "reason:empty_trace"
	tagReasonTraceIDZero           = "reason:trace_id_zero"
	tagReasonSpanIDZero            = "reason:span_id_zero"
	tagReasonForeignSpan           = "reason:foreign_span"
)
```

There is nothing wrong with being verbose if it helps with clarity. I find any of the above much easier to understand.

djova replied:
> Lastly, I think a good practice in Go would be to have all these tags as constants, for example:

That seems unnecessarily verbose. I think as long as the reason names are used only in one place then we don't need to do that? 

Following-up to the slack discussion we had I made several changes:

* replaced hacky JSON with `tagValues()` method on each struct, which returns a map of `reason_name -> count`. That map is then used for sending metrics and for generating the concise string representation for logs
* now we send counters for all stats even if count=0

gbbr replied:
Sure. SGTM

djova committed:
revert accidental Gopkg change

djova committed:
go fmt

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
-	// TracesDropped is the number of traces dropped.
-	TracesDropped int64
+	// TracesDropped contains stats about the count of dropped traces by reason
+	TracesDropped TracesDroppedStats
```
What I meant here was to use [type embedding](https://golang.org/doc/effective_go.html#embedding), like this:
```suggestion
	TracesDropped
```
This means calling the type `TracesDropped` (without the suffix), and as a result of embedding, the field as well.

djova replied:
Oooh I see. That's cool, had no clue this was possible. 

I just tried it out and I think it could hurt the readability of the code, because when incrementing some of these stats then you might not know whether the reason is related to dropped or malformed traces without jumping to where it's defined. 

Without the embedding it's clear:
* `atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)`
* `atomic.AddInt64(&ts.TracesMalformed.ServiceEmpty, 1)`

With the embedding it gets a bit confusing:
* `atomic.AddInt64(&ts.SpanIDZero, 1)`
* `atomic.AddInt64(&ts.ServiceEmpty, 1)`

So maybe better to keep the encapsulation unless there's another reason to use the embedding?

gbbr replied:
Oh but you don’t have to do the latter form. You can keep ts.TracesDropped.X even with embedding. I don’t think it works otherwise anyway, only for methods.

djova replied:
It actually works like that, all the fields become embedded as well. If we're just going to use the fully qualified `ts.TracesDropped.X` then what's the advantage of using the embedding?

gbbr replied:
I see. I honestly didn’t know that. I rarely use this feature. I just preferred the name without the suffix. You can go ahead and do whichever you prefer. Sorry to confuse

djova replied:
Oh ok. I'll rename to remove `Stats` suffix, and leaving it to use encapsulation instead of embedding.

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
+	// TracesDropped contains stats about the count of dropped traces by reason
+	TracesDropped TracesDroppedStats
+	// TracesMalformed contains stats about the count of malformed traces by reason
+	TracesMalformed TracesMalformedStats
```
Same here:
```suggestion
	TracesMalformed
```

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -86,22 +89,26 @@ func (rs *ReceiverStats) Reset() {
 
 // Strings gives a multi strings representation of the ReceiverStats struct.
-func (rs *ReceiverStats) Strings() []string {
+func (rs *ReceiverStats) Strings() (infoStrings []string, warnStrings []string) {
```
`Strings` is such a confusing name, given that `String` is so commonly used in Go as `fmt.Stringer` implementations. Maybe `Report` would be a better name.

gbbr replied:
Or even better, we could call it `Log` and have this very method call `log.Info` and `log.Warn`. It seems like it would be much cleaner. What do you think?

djova replied:
Good point, that is simpler. Refactored https://github.com/DataDog/datadog-agent/commit/6f0b21852341dcde2b3d6e12998695af640e5d22

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
+	return statsStructToMap(s.AtomicCopy(), "TracesDroppedStats")
+}
+
+func (s *TracesDroppedStats) String() string {
```
Here we can use `strings.Builder` to do something (again verbose) like:
```go
func (s *TracesDroppedStats) String() string {
    var str strings.Builder
    for tag, value := range map[string]int64{
        tagReasonDecodingError: atomic.LoadInt64(&s.DecodingError),
        tagReasonEmptyTrace: atomic.LoadInt64(&s.EmptyTrace),
        // ...
    } {
        if value > 0 {
            str.WriteString(tag)
            str.WriteByte(':')
            str.WriteString(strconv.Itoa(valut))
            str.WriteByte(' ')
        }
    }
    return str
}
```
This is again super-verbose, but IMO much clearer. Generally, this is the preferred approach in Go. It might seem odd coming from other languages.

djova replied:
I'm not convinced that it's clearer. Main problem with this approach is that you have to take care not to add an extra separator at the end. Since we're using the comma as a separator to match with the rest of the formatting we'll end up with an extra comma then. 

Isn't `strings.Join(listOfStrings, ", ")` pretty clear? It does basically the same thing inside using a builder, and you know it'll handle placing the separator edge cases correctly.

gbbr replied:
Sure. You can use the join. Performance isn’t critical here. It was just an example. Creating a slice and then joining will likely be the easiest.

gbbr reviewed (commented):
Added some suggestions which will help remove the magic. In general, I think it may be possible to find an easier, more automated way (using [custom struct tags and the `reflect` package](https://golang.org/pkg/reflect/#StructTag) or by using [code generation](https://blog.golang.org/generate)), but that requires implementing this automation and applying it throughout, for consistency.

I personally think it's not worth the trouble and being verbose here is not such a big sacrifice. This package in itself is already being very verbose. If we are to apply a new pattern here, we should apply it throughout to be consistent. But I think that's a bigger job, possibly outside the scope of this PR and not possible within the timeframe that you were given.

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -4,8 +4,10 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
```
This import needs to be moved down, outside of the "standard library" group.

djova replied:
Oh weird not sure why my editor put it there. Need to fix the auto-import management. Updated

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -32,37 +35,74 @@ var (
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		log.Debugf("Dropping invalid trace (reason:trace_id_zero): %s", s)
+		return errors.New("trace_id_zero")
```
A better way to do this is:
```go
return fmt.Errorf("Dropping invalid trace (reason:trace_id_zero): %s", s)
```
And not logging here. We should log where the error is returned, at the very call of `normalizeTrace`. Then, this makes the errors more useful, rather than just a filler with a random name.

djova replied:
Good point, updated

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -392,10 +387,16 @@ func (r *HTTPReceiver) loop() {
 				info.UpdateReceiverStats(accStats)
 
-				for _, logStr := range accStats.Strings() {
+				infoStrings, warnStrings := accStats.Strings()
```
It's saner to just call `accStats.Log()` here and that's all. All this logic seems unnecessary. Then, `accStats.Log()` can take care of everything.

djova replied:
Done

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -255,14 +417,29 @@ func (s *Stats) String() string {
 
+// WarnString returns a string representation of the Stats struct containing only issues which we should be warning on
+// if there are no issues then an empty string is returned
+func (ts *TagStats) WarnString() string {
```
Can we do this in `accStats.Log`?

djova replied:
We could, but then we would have to move both of these in there:

* `func (s *Stats) InfoString() string`
* `func (ts *TagStats) WarnString() string`

Which would probably make the `accStats.Log` too cluttered?

gbbr replied:
I don't think so, but I don't feel strongly about it. So if you're convinced that this is better, then leave it. My thought it simply that these aren't used elsewhere. And it's weird why they are called `WarnString` and `InfoString` when the implementation has nothing to do with "warn" and "info". It's more like what they do.

At the very least, it would be nice to lower-case them so that they are not exported as part of the API of this package.

gbbr reviewed (commented):
Sorry for the multiple separate groups of comments.

djova committed:
add release notes

djova committed:
Revert "add release notes"

Wrong branch.

This reverts commit 71eece86069bb16c175ab2f20df2b715805d260a.

djova committed:
remove hacky json marshaling

djova committed:
rename toMap, cleanup

djova committed:
refactor stats logging

djova committed:
simplify normalize, foreign check outside

djova committed:
fmt

djova committed:
fix imports

djova committed:
rename

djova committed:
refactor

djova committed:
comment

djova committed:
comment fix

djova committed:
revert debug code

djova committed:
trace-agent: more permissive trace normalization with metrics/logs

**Malformed traces** - the trace agent is now more permissive with malformed traces, keeping track of the number of malformed traces in a new `datadog.trace_agent.receiver.traces_malformed` metric (tagged by `reason`):

* The following fields are now replaced with reasonable defaults if invalid or empty and truncated if exceeding max length: `span.service`, `span.name`, `span.resource`, `span.type`.
* `span.duration=0` is now allowed.
* Missing span start date now defaults to current time.

**Dropped traces** - we now drop traces only for a couple of reasons: `empty_trace` or `foreign_span`.

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,81 @@ const (
-	// tolerate for span end dates
-	MaxEndDateOffset = 10 * time.Minute
+	// DefaultServiceName TODO: do we really want such indescriptive defaults here like "service"?
+	DefaultServiceName = "no_name_service"
```
@gbbr @palazzem do we really want such generic names as defaults or should it be something like this to make it more obvious in the UI that the name is missing?

djova replied:
True we already agreed on it, and these values will probably be fine. Updated to use them https://github.com/DataDog/datadog-agent/pull/3738/commits/872d3d410a1a02efb0f92cce7b7c0dbd3f52919e

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
-	// TracesDropped is the number of traces dropped.
-	TracesDropped int64
+	// TracesDropped contains stats about the count of dropped traces by reason
+	TracesDropped TracesDropped
```
Let's make them (`Traces{Dropped,Malformed}`) pointers, so we don't create copies every time we reference them.

djova replied:
Are you sure it creates copies? If `ts.TracesDropped` created a copy every time you accessed it, then `atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)` shouldn't work because it would be incrementing a copy?

gbbr replied:
That’s because ts is a pointer, and adding the & sign takes the address of the field. The copy happens on assignment. It’s better to copy a pointer than a struct. Not a problem now but I think better be safe?

djova replied:
Yeah makes sense. Updated to use pointers.

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+
+// inlineNonZeroTagValues serializes the entries in this map into format "key1: value1, key2: value2, ...", sorted by
+// key to ensure consistent output order. Only non-zero values are included.
+func inlineNonZeroTagValues(statsMap map[string]int64) string {
```
I've learned and felt that it's better to name functions based on what they do, not how or where they are used. This function takes a map and returns a string representation of it. Can we call it as such? Maybe `mapString` or `mapToString` ?
```go
// mapString returns a string representation of the given map, eluding empty values.
func mapToString(m map[string]int64) string {
```

djova replied:
Sure, updated.

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+// inlineNonZeroTagValues serializes the entries in this map into format "key1: value1, key2: value2, ...", sorted by
+// key to ensure consistent output order. Only non-zero values are included.
+func inlineNonZeroTagValues(statsMap map[string]int64) string {
+	keys := make([]string, len(statsMap))
```
You can drop the `i` if you do:
```go
keys := make([]string, 0, len(statsMap))
for k := range statsMap {
    keys = append(keys, k)
}
```

djova replied:
Oh cool, updated

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+			results = append(results, fmt.Sprintf("%s:%d", key, value))
+		}
+	}
+
```
Would you mind removing this empty line? I feel that the closure of the loop above it already creates enough spacing.

djova replied:
Removed

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+
+// TracesMalformed contains counts for reasons malformed traces have been accepted after applying automatic fixes
+type TracesMalformed struct {
+	DuplicateSpanID       int64
```
Maybe we should document these, but I don't feel strongly about it. Just a thought...

djova replied:
Yeah probably good to document. Added docs.

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,46 @@
+	"testing"
+)
+
+func TestTracesDroppedStatsToMap(t *testing.T) {
```
Here it's fine to create 2 tests with 2 sub-tests each:
* `TestTracesDropped` with sub-tests `StatsToMap` and `StatsToString`
* `TestTracesMalformed` with sub-tests `StatsToMap` and `StatsToString`

djova replied:
Yeah I guess that's a bit nicer here. Updated.

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -322,10 +374,10 @@ func TestNormalizeInvalidUTF8(t *testing.T) {
 		assert := assert.New(t)
 
-		span := testSpan()
+		ts, span := &info.TagStats{}, newTestSpan()
```
Can we please not do this and instead just:
```go
ts := &info.TagStats()
span := newTestSpan()
```

djova replied:
Sure, updated

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +36,326 @@ func testSpan() *pb.Span {
-	span1 := testSpan()
-
-	span2 := testSpan()
+	ts, span1, span2 := &info.TagStats{}, newTestSpan(), newTestSpan()
```
Same here and everywhere else please 🙏. Declaring variables like this is a bit confusing. We're not really doing it anywhere. As far as I've learned, the only idiomatic and valid reasons to do this, in general, are:
* for loops with multiple variables
* if statements
```go
if n1, n2 := len(a), len(b); n1 != n2 {
    // ...
}
```
* swaps:
```go
a, b = b, a
```
I'm sure some rare case can be found where it makes sense within declarations too, but I haven't found it and this doesn't look like one of them.

djova replied:
Ok, updated. I figured it was simpler to do it all one one line, but fine like this as well.

gbbr reviewed (commented):
Thanks Dusan. This is looking much better. Left some (more) nits.

djova committed:
use previously agreed-upon defaults

djova committed:
remove comment

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,81 @@ const (
+		fallbackServiceName = ts.Lang
+	}
+
+	// While it's technically possible to get TraceID=0 (with 1 in 2**64 probability) we figured it's far more likely
```
Good point, removed.

djova committed:
moving dropped trace error logging

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -134,31 +171,31 @@ func normalize(s *pb.Span) error {
 	if len(t) == 0 {
-		return errors.New("empty trace")
+		atomic.AddInt64(&ts.TracesDropped.EmptyTrace, 1)
+		return errors.New("dropping invalid trace (reason:empty_trace)")
```
Good point, updated

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -134,31 +171,31 @@ func normalize(s *pb.Span) error {
-			return errors.New("empty `TraceID`")
+		if span.TraceID != firstSpan.TraceID {
+			atomic.AddInt64(&ts.TracesDropped.ForeignSpan, 1)
+			return fmt.Errorf("dropping invalid trace (reason:foreign_span): %s", span)
```
I updated to remove the "dropping invalid trace" language to where it's actually dropped, but I think we should keep the full span there, as we're logging the full span in every other malformed trace case as well. These logs are only going to show up if someone enables debug logging, and if they enable debug logging with a goal of figuring out why they have dropped/malformed traces, then it's probably most useful to print the whole span out?

gbbr replied:
It could be. Sometimes we had reports of it as being spammy, but since it is only for `DEBUG` level I guess it's fine. 👍

djova committed:
import sort

djova commented on pkg/trace/info/stats.go:
```diff
@@ -2,7 +2,9 @@ package info
 
 import (
 	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
```
Ah again. I've updated my editor to use `goimports` sorting so this should hopefully be fixed automatically going forward. We should probably have `goimports` sorting be part of the build so this is enforced automatically?

gbbr replied:
I think this is something related to your editor. I don't have this problem with `vim-go`.

djova committed:
remove dropping language from error

djova committed:
parens for consistency

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,81 @@ const (
+	// consequences. Same thing for SpanID.
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		return fmt.Errorf("dropping invalid trace (reason:trace_id_zero): %s", s)
```
Removed the `dropping invalid trace` language as that's now done where the trace is dropped, but leaving the reason-tag like formatting to have it stay consistent with all the other logs, the metrics, and how the reasons are reported in `datadog-agent status`.

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,81 @@ const (
+	}
+	if s.SpanID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)
+		return fmt.Errorf("dropping invalid trace (reason:span_id_zero): %s", s)
```
Updated in same way as mentioned here https://github.com/DataDog/datadog-agent/pull/3738#discussion_r297533505

gbbr replied:
Sure. I guess the reason names make sense to a user too.

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,81 @@ const (
 	if s.Name == "" {
-		return errors.New("empty `Name`")
+		atomic.AddInt64(&ts.TracesMalformed.SpanNameEmpty, 1)
+		log.Debugf("Fixing malformed trace (reason:span_name_empty), setting span.name=%s: %s", DefaultSpanName, s)
```
Looks like it does, here's how it gets formatted as `%s`: `Fixing malformed trace (reason:service_empty), setting span.service=service: name:"django.controller" resource:"GET /some/raclette" traceID:424242 spanID:2382721542312139272 parentID:1111 start:1448466874000000000 duration:10000000 meta:<key:"pool" value:"fondue" > meta:<key:"user" value:"leo" > metrics:<key:"cheese_weight" value:100000 > type:"http"`


And `%#v`: `Fixing malformed trace (reason:service_empty), setting span.service=service: &pb.Span{Service:"", Name:"django.controller", Resource:"GET /some/raclette", TraceID:0x67932, SpanID:0xb4e82abb8e515c1f, ParentID:0x457, Start:1448466874000000000, Duration:10000000, Error:0, Meta:map[string]string{"pool":"fondue", "user":"leo"}, Metrics:map[string]float64{"cheese_weight":100000}, Type:"http"}`

Changed to `%#v` though as that is a bit clearer since the whole span is wrapped in `&pb.Span{...}`.

gbbr replied:
Actually no, let's leave it with `%s` as that's perfect. It was designed to be user friendly. `#%v` will only make sense to Go users, while `%s` is more language agnostic.

djova replied:
Reverted

djova committed:
use Go-syntax representation for printing spans

djova commented on pkg/trace/api/api.go:
```diff
@@ -308,17 +308,10 @@ func (r *HTTPReceiver) processTraces(ts *info.TagStats, traces pb.Traces) {
+		err := normalizeTrace(ts, trace)
 		if err != nil {
-			atomic.AddInt64(&ts.TracesDropped, 1)
+			log.Debug(err)
```
Updated

djova committed:
formatting

djova commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+	for reason, count := range ts.TracesDropped.tagValues() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_dropped", count, append(tags, "reason:"+reason), 1)
+	}
+
```
Removed

djova committed:
private methods

djova committed:
Revert "use Go-syntax representation for printing spans"

This reverts commit 1d652d2ca2860847ce1c0e03f6933b7968c5fb10.

djova committed:
smaller variable names

djova commented on pkg/trace/info/stats.go:
```diff
@@ -255,14 +396,29 @@ func (s *Stats) String() string {
+// if there are no issues then an empty string is returned
+func (ts *TagStats) WarnString() string {
+	var warnings []string
+	droppedReasons := ts.TracesDropped.String()
```
Ok, updated

djova commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +36,326 @@ func testSpan() *pb.Span {
-	assert.NoError(t, normalize(s))
+	ts, s := &info.TagStats{}, newTestSpan()
+	assert.NoError(t, normalize(ts, s))
+	assertNoNormalizationIssues(t, ts)
```
Yes, but if there is an unexpected normalization issue then the log message won't be as descriptive. As it is now it will say `found unexpected normalization issue %s=%d`, with the reason matching the reason names used everywhere else. So I think better leave it like that. And it's only a couple of lines of code for `assertNoNormalizationIssues` so it's not like switching to that would make things significantly simpler.

djova replied:
Updated as explained in https://github.com/DataDog/datadog-agent/pull/3738#discussion_r297788738

djova committed:
use assert.Equal with map

djova commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +36,326 @@ func testSpan() *pb.Span {
+	s.Service = ""
+	assert.NoError(t, normalize(ts, s))
+	assert.Equal(t, s.Service, DefaultServiceName)
+	assertNormalizationIssue(t, ts, "service_empty")
```
That's cool, didn't realize the tests could compare structs in detail. 

I think the current approach is better for several reasons:

* it's testing the exact reason names `service_empty` used in tags, logs, and agent status instead of the internal counters `TracesMalformed.ServiceEmpty`. It's better to test that because we're testing the "interface" and not the underlying implementation. If we were testing only the counters we might not catch that a tag name was wrong. 
* `assertNormalizationIssue(t, ts, "service_empty")` is a bit easier to read than `assert.Equal(t, ts, tagStatsMalformed(&info.TracesMalformed{ServiceEmpty: 1})`
* broken test output is a bit clearer since we're only comparing the present issues and it won't show a lot of other unrelated spam like all the zero values when it does `%#v` formatting of the entire structs

But I like the idea of comparing objects directly. Updated to use direct comparison of maps instead of hacky manual diff https://github.com/DataDog/datadog-agent/pull/3738/commits/86d706ec95d3df9f5f833deb1a1a2eee8080ea68 what do you think of that?

gbbr replied:
> `assertNormalizationIssue(t, ts, "service_empty")` is a bit easier to read than `assert.Equal(t, ts, tagStatsMalformed(&info.TracesMalformed{ServiceEmpty: 1})`

I disagree. Nobody knows what `assertNormalizationIssue` unless they go to a definition and learn a new function. The latter is already known.

> broken test output is a bit clearer since we're only comparing the present issues and it won't show a lot of other unrelated spam like all the zero values when it does %#v formatting of the entire structs

Have you tried it? The output is pretty good.

djova replied:
> Have you tried it? The output is pretty good.

You're right,  the diff output is actually fine, so it's good to use the object comparison. 

I updated to have all asserts done right in the tests, not in these other functions. Still using a bit of code to simplify the tests, but I think that's fairly reasonable. 

It now looks like this:

Asserting no issues: `assert.Equal(t, noIssues(), statsToIssues(ts))`
Asserting a specific issue: `assert.Equal(t, onlyIssue("span_name_empty"), statsToIssues(ts))`

I think this should be ok?

djova committed:
rename

djova committed:
simplify

djova committed:
remove line

djova committed:
subtests

djova committed:
separate lines

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -114,10 +119,18 @@ func newTagStats(tags Tags) *TagStats {
 	return &TagStats{tags, Stats{}}
 }
 
+// AllNormalizationIssues returns a map of counts of all normalization issue reasons due to dropped or malformed traces
```
Let's remove this. It's not used anymore.

djova replied:
Removed

djova committed:
add docs for reasons

djova committed:
simplify tests

djova committed:
TagStats pointers for internal structs

djova committed:
cleanup

djova committed:
fmt

djova commented:
@gbbr could you take another look? I think I've addressed all of the comments.

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		return fmt.Errorf("(reason:trace_id_zero): %s", s)
```
These error messages should be user friendly and good to be used anywhere. They should also be human readable. The way they are written now they seem like they are expected to be appended to another string, which they are, but they should work on their own. Please make all of them human friendly:
```suggestion
		return fmt.Errorf("TraceID is zero: %s", s)
```

djova replied:
Wouldn't it be better to maintain consistency with the metric tags, log warning output, and `datadog-agent status` output? The main reason anyone would use these logs is if they saw either the dropped/malformed trace metrics, or saw the periodic summary log message, in which case isn't it more user friendly to maintain the exact same strings? I imagine most people would do a grep for `trace_id_zero` to find the right log message? 

As for the `(reason:trace_id_zero)` formatting, the goal is to maintain a consistent format forall of these trace normalization debug log messages:

*`Fixing malformed trace (reason:invalid_http_status_code) ...`
*`Dropping invalid trace (reason:trace_id_zero) ...`

That's why initially I figured it's simplest to do the logging right where the decision is being made. I moved it out though so the error now seems a bit out of place, but the log messages will at least be consistent.

gbbr replied:
Why does logging need to be consistent with tags used in stats? That's the part I don't get. These logs are mostly for humans, stats are written in a more cryptic fashion due to their limitations. To me `fmt.Errorf("(reason:trace_id_zero): %s", s)` does not look like a nice error. The error should make sense on its own and shouldn't make assumptions about the context.

djova replied:
> Why does logging need to be consistent with tags used in stats? That's the part I don't get. These logs are mostly for humans, stats are written in a more cryptic fashion due to their limitations.

That is precisely why it needs to match. Because this is what a *human* is going to experience:

1. sees this in the periodic summary logs: `WARN: malformed_traces(empty_span_name: 4, invalid_duration:3) Enable debug logging for more information` 
1. enables debug logging
1. Greps logs for `empty_span_name` or `invalid_duration`, finds more inforamtion
1. Goes to metrics to plot when this started using tags `reason:empty_span_name`, `reason:invalid_duration`
1. ...

It's going to make the human's life much easier if all of that is consistent. 

> To me `fmt.Errorf("(reason:trace_id_zero): %s", s)` does not look like a nice error. The error should make sense on its own and shouldn't make assumptions about the context.

Ok, I updated the error and logging to stand better on its own Alright, updated here https://github.com/DataDog/datadog-agent/pull/3738/commits/6ff5cf52cc4ce0b718cbdac1a3f4a094e80abcd0

However, I really think we should keep the reason names consistent here with the rest of the logging and metrics.

gbbr replied:
Ok, let's make a mixture of both then, for example:
```go
return fmt.Errorf("TraceID is zero (reason:trace_id_zero)")
```
Would you be happy with that? The same for all the others.

djova replied:
Do we really need that though? It's adding unnecessary duplication and making it more verbose for no real benefit:

* `Trace is Empty (reason:empty_trace)`
* `Trace has foreign span (reason:foreign_span)`
* `Trace has duplicate span ID (reason:duplicate_span_id)`
* `Trace has invalid service (reason:service_invalid)`
* ...

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+	}
+	if s.SpanID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)
+		return fmt.Errorf("(reason:span_id_zero): %s", s)
```
Same here:
```suggestion
		return fmt.Errorf("SpanID is zero: %s", s)
```

djova replied:
Same issue as https://github.com/DataDog/datadog-agent/pull/3738#discussion_r298042193, resolving to maintain one discussion.

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -134,31 +168,31 @@ func normalize(s *pb.Span) error {
 	if len(t) == 0 {
-		return errors.New("empty trace")
+		atomic.AddInt64(&ts.TracesDropped.EmptyTrace, 1)
+		return errors.New("(reason:empty_trace)")
```
```suggestion
		return errors.New("empty trace")
```

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -134,31 +168,31 @@ func normalize(s *pb.Span) error {
-			return errors.New("empty `TraceID`")
+		if span.TraceID != firstSpan.TraceID {
+			atomic.AddInt64(&ts.TracesDropped.ForeignSpan, 1)
+			return fmt.Errorf("(reason:foreign_span): %s", span)
```
```suggestion
			return fmt.Errorf("foreign span in trace (%d != %d): %s", span.SpanID, firstSpan.TraceID, span)
```
Let's make them all crystal clear for a human. Error logs don't need to be consistent with stats terminology.

djova replied:
Same issue as https://github.com/DataDog/datadog-agent/pull/3738#discussion_r298042193, resolving to maintain one discussion.

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +34,358 @@ func testSpan() *pb.Span {
+// statsToIssues returns map of {reason->count} for normalization issues due to dropped or malformed traces
+func statsToIssues(ts *info.TagStats) map[string]int64 {
+	m := ts.TracesDropped.TagValues()
+	for r, c := range ts.TracesMalformed.TagValues() {
```
I think this is still not the right approach. The fact that you had to make `TagValues` exported and part of the public API of the `info` package only for it to be used in the tests of another package, is indicative of design problems. The tests we write should not span across testing multiple packages at once and be a hybrid of unit and integration tests. It should either be one or the other.

* `tagValues` converts structure fields to a map. If you want to guarantee that all fields are mapped correctly to the right strings (reasons), that should be done within the `info` package, not here.
* the tests in this package are a mix of an integration and unit test: they don't test that the normalizer correctly update the stats; they test that the normalizer updates the stats _and also_ maps itself to the right key. Do you see the problem here? We should only do the former. The fact this implementation is providing difficulties is a good indication that it is not the right approach.

I would like to kindly ask you to go back to [this suggestion](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r297626197) which is a very clear, contained unit test free of any additional helpers and unnecessary exports. The last part that your tests cover for (that fields map to the right map keys aka. reasons) does not belong in this package, it belongs in the `info` package. The fact the `tagValues` had to become exported only for test purposes is a clear indication of it. There is no reason to complicate things.

If you really wish to write an integration test, you can do so by using the [test package](https://godoc.org/github.com/DataDog/datadog-agent/pkg/trace/test), with some examples found in the [documentation](https://godoc.org/github.com/DataDog/datadog-agent/pkg/trace/test) and real world ones in the [testsuite](https://github.com/DataDog/datadog-agent/tree/master/pkg/trace/test/testsuite) folder. That will allow you to programmatically spin up a trace-agent, a fake backend, send data to it and assert what the backend receives. Then, you can send malformed or bad traces and assert that they don't reach the server or that they do so given the correct transformation.

But please refrain from creating a mix of unit and integration tests. Let's stay purely unit focused. If we test that the normalizer sets the correct stats in this package, and we test that the stats get mapped to the expected keys in the other package, then it is fine to conclude that they will also work together.

If you want to write integrations tests, then please do that as part of the `test` package. I'd say it's quite easy to use and beneficial. I'm happy to help use it if anything is confusing.

djova replied:
> The fact that you had to make `TagValues` exported and part of the public API of the `info` package only for it to be used in the tests of another package, is indicative of design problems.

Good point. Ok, updated to compare the structs directly. I don't think we need another integration test on top of this, that may be overkill. This should be ok.

gbbr replied:
I also think it's ok, but I wanted to let you know about it, in case we want to use it. You can even inspect the logs that the agent outputs.

gbbr reviewed (changes requested):
Thanks Dusan! This is looking real nice now. I just have a few more suggestions around logging and tests. Almost there. 😄

gbbr commented:
We forgot to take into account decoding errors, we'll have to add this somewhere above or below [this line](https://github.com/DataDog/datadog-agent/blob/1f8b824dbe824c1fea217ae3ef3ff4e25f382b05/pkg/trace/api/api.go#L510):

```go
if count, err := strconv.Atoi(req.Header.Get(sampler.TraceCountHeader)); err == nil {
    atomic.AddInt64(&ts.TracesDropped.DecodingError, count)
} else {
    metrics.Count("datadog.trace_agent.receiver.decoding_error", 1, nil, 1)
}
```

djova committed:
tests compare structs, not tag values

djova committed:
track decoding error

djova committed:
fmt

djova commented:
> We forgot to take into account decoding errors, we'll have to add this somewhere above or below [this line](https://github.com/DataDog/datadog-agent/blob/1f8b824dbe824c1fea217ae3ef3ff4e25f382b05/pkg/trace/api/api.go#L510):

Good catch! Totally missed that one. Added https://github.com/DataDog/datadog-agent/pull/3738/commits/8774469ea836509cda073f3d45d38f98fc4b8467

djova committed:
update dropped trace error and logging

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +268,50 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 
+	traces, err := getTraces(v, w, req)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "Cannot decode") {
```
This is not how we check errors. If you really want to change `getTraces` to return errors, you'll need to create a set of custom errors. That's the idiomatic way to do it in Go, but it will be tedious. I'll try to illustrate. 
```go
type decodingError struct {err error}

func (e *decodingError) Error() string { return fmt.Sprintf("decoding error: %s", e.err) }

type formatError struct {t string}

func (e *formatError) Error() string { return fmt.Sprintf("unsupported media type: %s", e.t) }

type versionError struct {v Version}

func (e *versionError) Error() string { return fmt.Sprintf("unsupported version: %s", e.v) }
```
Then, you would here do:
```go
traces, err := getTraces(v, req)
switch e := err.(type) {
    case decodingError:
        if traceCount > 0 {
            atomic.AddInt64(&ts.TracesDropped.DecodingError, traceCount)
        } else {
            metrics.Count("datadog.trace_agent.receiver.decoding_error", 1, nil, 1)
        }
        log.Errorf("Cannot decode %s traces payload: %v", v, err)
        httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)

    case formatError:
        httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))

    case versionError:
        httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
}
```
This would significantly clean up `getTraces` to the extent that it's equally long as the above switch statement, making this whole suggestion overkill, even if more correct. What I would like to suggest instead is to either:
* Inline all of `getTraces` into `handleTraces` right here. It doesn't make much sense to have it as a separate function to be honest, or,
* Just use the change I initially suggested [here](https://github.com/DataDog/datadog-agent/pull/3738#issuecomment-506373265) and nothing else.

I am also ok with the overkill example above, but frankly I think it's a bit too much. The bottom line is that we shouldn't mix returning errors with handling or logging those very same errors. It also seems weird that a function called `getTraces` is replying to the HTTP client. Personally I think it'd serve as a nice cleanup to inline `getTraces` into `handleTraces`

gbbr replied:
I tried it out:
```diff
diff --git a/pkg/trace/api/api.go b/pkg/trace/api/api.go
index aeb02b13..06235a82 100644
--- a/pkg/trace/api/api.go
+++ b/pkg/trace/api/api.go
@@ -273,8 +273,32 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 		return
 	}
 
-	traces, ok := getTraces(v, w, req)
-	if !ok {
+	var traces pb.Traces
+	mediaType := getMediaType(req)
+	switch v {
+	case v01:
+		switch mediaType {
+		case "application/json", "text/json", "":
+			// ok
+		default:
+			httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))
+			return
+		}
+		var spans []pb.Span
+		if err := json.NewDecoder(req.Body).Decode(&spans); err != nil {
+			log.Errorf("Cannot decode %s traces payload: %v", v, err)
+			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
+			return
+		}
+		traces = tracesFromSpans(spans)
+	case v02, v03, v04:
+		if err := decodeReceiverPayload(req.Body, &traces, v, mediaType); err != nil {
+			log.Errorf("Cannot decode %s traces payload: %v", v, err)
+			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
+			return
+		}
+	default:
+		httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
 		return
 	}
 
@@ -479,48 +503,6 @@ func (r *HTTPReceiver) Languages() string {
 	return strings.Join(str, "|")
 }
 
-func getTraces(v Version, w http.ResponseWriter, req *http.Request) (pb.Traces, bool) {
-	var traces pb.Traces
-	mediaType := getMediaType(req)
-	switch v {
-	case v01:
-		// We cannot use decodeReceiverPayload because []model.Span does not
-		// implement msgp.Decodable. This hack can be removed once we
-		// drop v01 support.
-		switch mediaType {
-		case "application/json", "text/json", "":
-			// ok
-		default:
-			httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))
-			return nil, false
-		}
-
-		// in v01 we actually get spans that we have to transform in traces
-		var spans []pb.Span
-		if err := json.NewDecoder(req.Body).Decode(&spans); err != nil {
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
-			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-			return nil, false
-		}
-		traces = tracesFromSpans(spans)
-	case v02:
-		fallthrough
-	case v03:
-		fallthrough
-	case v04:
-		if err := decodeReceiverPayload(req.Body, &traces, v, mediaType); err != nil {
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
-			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-			return nil, false
-		}
-	default:
-		httpEndpointNotSupported([]string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-		return nil, false
-	}
-
-	return traces, true
-}
-
 func decodeReceiverPayload(r io.Reader, dest msgp.Decodable, v Version, mediaType string) error {
 	switch mediaType {
 	case "application/msgpack":
```
I feel this would work fine without all the weird functions.

djova replied:
Good points. Yeah better not make this too complicated by returning errors and doing checking outside. 

Updated to move everything into `getTraces`.

gbbr reviewed (commented):
Thanks for the intention of improving the code. It's much appreciated. I left some suggestions.

djova committed:
private

djova commented on pkg/trace/api/api.go:
```diff
@@ -37,6 +38,10 @@ const (
+
+	// TraceCountHeader is the header client implementation should fill
+	// with the number of traces contained in the payload.
+	TraceCountHeader = "X-Datadog-Trace-Count"
```
Ok, updated

djova committed:
traceCount refactor

djova commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +268,50 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 	}
 }
 
+func getTraceCount(req *http.Request) (traceCount int64) {
```
Sure, updated

djova committed:
cleanup

djova commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +268,50 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 }
 
+func getTraceCount(req *http.Request) (traceCount int64) {
+	if traceCountStr := req.Header.Get(TraceCountHeader); traceCountStr != "" {
```
Yeah that's a bit cleaner, updated

djova committed:
cleanup getTraces

djova commented on pkg/trace/api/api.go:
```diff
@@ -491,16 +507,16 @@ func getTraces(v Version, w http.ResponseWriter, req *http.Request) (pb.Traces,
-			httpFormatError(w, v, fmt.Errorf("unsupported media type: %q", mediaType))
-			return nil, false
+			err := fmt.Errorf("unsupported media type: %q", mediaType)
+			httpFormatError(w, v, err)
```
Yeah good point, updated to how it was before, return bool.

djova commented on pkg/trace/api/api.go:
```diff
@@ -491,16 +507,16 @@ func getTraces(v Version, w http.ResponseWriter, req *http.Request) (pb.Traces,
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
 			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
-			return nil, false
+			return nil, log.Errorf("Cannot decode %s traces payload: %v", v, err)
```
Updated

djova commented on pkg/trace/api/api.go:
```diff
@@ -498,8 +507,9 @@ func getTraces(v Version, w http.ResponseWriter, req *http.Request) (pb.Traces,
 		if err := json.NewDecoder(req.Body).Decode(&spans); err != nil {
-			log.Errorf("Cannot decode %s traces payload: %v", v, err)
 			httpDecodingError(err, []string{tagTraceHandler, fmt.Sprintf("v:%s", v)}, w)
+			atomic.AddInt64(&ts.TracesDropped.DecodingError, traceCount)
```
Note that we don't actually need to report `metrics.Count("datadog.trace_agent.receiver.decoding_error", 1, nil, 1)` here as it turns out `httpDecodingError` is already reporting a metric `datadog.trace_agent.receiver.error` tagged by `error:decoding-error`. So in the case of `traceCount=0` for a missing trace count header, we still have that metric to track the number of decoding errors.

gbbr replied:
True 👍 We should remove it and possibly update our RFC.

djova replied:
Ok, renamed "traces malformed" to "spans malformed" and updated RFC.

djova committed:
getTraces inline

djova committed:
rename traces_malformed to spans_malformed

gbbr commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		return fmt.Errorf("trace_id_zero: %s", s)
```
I couldn't comment anymore [here](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r298679652) for some reason, but we should either have (here and everywhere else):
```go
return fmt.Errorf("TraceID is zero")
```
or
```go
return fmt.Errorf("TraceID is zero (reason:trace_id_zero)")
```
If you really want to have the tag name there.

But not what we have now please. And errors, as opposed to logs should all be lower-cased (except the above example which represents a span field).

djova replied:
Alright... more prose it is https://github.com/DataDog/datadog-agent/pull/3738/commits/9767bc2835f66e290df19b09c3c1ec55d52b11ff

djova committed:
more prose in logging messages

djova committed:
update log

djova committed:
reverse words to  match metric name

djova committed:
add missing modification check

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -155,15 +158,122 @@ func (ts *TagStats) publish() {
+
+// mapToString serializes the entries in this map into format "key1: value1, key2: value2, ...", sorted by
+// key to ensure consistent output order. Only non-zero values are included.
+func mapToString(statsMap map[string]int64) string {
```
```suggestion
func mapToString(m map[string]int64) string {
```
It's just a `map`, not a "stats map" :)

djova replied:
Updated

gbbr commented on pkg/trace/info/stats.go:
```diff
@@ -198,7 +308,28 @@ type Stats struct {
 	atomic.AddInt64(&s.TracesReceived, atomic.LoadInt64(&recent.TracesReceived))
-	atomic.AddInt64(&s.TracesDropped, atomic.LoadInt64(&recent.TracesDropped))
+
+	// TracesDropped
```
Do we really need these comments? Do they add anything?

djova replied:
Removed. Was just following comment style I saw elsewhere in this code

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+func TestTracesDropped(t *testing.T) {
+	var s TracesDropped
```
This is funky, let's be clear here:
```go
s := TracesDropped{
    DecodingError: 1,
    ForeignSpan: 1,
}
```

djova replied:
Good point, updated

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func TestSpansMalformed(t *testing.T) {
+	var s SpansMalformed
```
Same here.

djova replied:
Updated

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,47 @@
+	})
+
+	t.Run("StatsToString", func(t *testing.T) {
+		assert.Equal(t, "decoding_error:1, foreign_span:1", s.String())
```
This test is weird. You could've used a [for loop with multiple test cases](https://github.com/golang/go/wiki/TableDrivenTests), having the input be a `TracesDropped` structure, asserting against a string output. This way, you're only testing two keys out of all.

djova replied:
It's just a sanity check that the function works, and that the keys are sorted correctly. We could enumerate all the test cases but that's probably overkill. I added a couple more keys, but I don't think enumerating lots of possible combinations of keys here is going to make the test any more useful.

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,47 @@
+	s.DecodingError++
+	s.ForeignSpan++
+
+	t.Run("StatsToMap", func(t *testing.T) {
```
```suggestion
	t.Run("tagValues", func(t *testing.T) {
```

djova replied:
Updated

gbbr commented on pkg/trace/info/stats_test.go:
```diff
@@ -0,0 +1,47 @@
+		}
+	})
+
+	t.Run("StatsToString", func(t *testing.T) {
```
```suggestion
	t.Run("String", func(t *testing.T) {
```

djova replied:
Updated

gbbr commented on releasenotes/notes/more-permissive-trace-normalization-20ed9bc015098341.yaml:
```diff
@@ -0,0 +1,17 @@
+---
+enhancements:
+  - |
+    APM: the trace agent now lets through a wider variety of traces, automatically correcting some malformed traces
```
👍 Nice

gbbr commented on pkg/trace/api/normalizer_test.go:
```diff
@@ -33,277 +34,350 @@ func testSpan() *pb.Span {
+	before := s.Start
+	assert.NoError(t, normalize(ts, s))
+	assert.Equal(t, tsMalformed(&info.SpansMalformed{InvalidStartDate: 1}), ts)
+	assert.True(t, s.Start > before, "start should have been reset to current time")
```
You can use https://golang.org/pkg/time/#Time.Before

djova replied:
But then need to construct `time.Time` struct, where we're only working with `int64` here. Maybe simplest just to leave it like this?

gbbr replied:
Yep, you're right. I missed that. It's fine to leave it like this. 👍

gbbr reviewed (commented):
I think this is probably the last round of comments :)

djova commented on pkg/collector/corechecks/embed/apm_nix.go:
```diff
@@ -24,8 +24,9 @@ const apm_binary_name = "trace-agent"
 		return binPath, nil
 	}
-	return binPath, fmt.Errorf("Can't access the default apm binary at %s", binPath)
+	return binPath, fmt.Errorf("Can't access the default apm binary at %s: %s", binPath, err.Error())
```
Oh committed this accidentally. I was debugging trying to run the check locally. Might as well make this change anyway, seems useful to have the error.

Commit:
Update pkg/collector/corechecks/embed/apm_nix.go

Co-Authored-By: Gabriel Aszalos <gabriel.aszalos@datadoghq.com>

djova commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +267,75 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
+	mediaType := getMediaType(req)
+	switch v {
+	case v01:
+		// We cannot use decodeReceiverPayload because []model.Span does not
```
Removed

djova commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +267,75 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
+			return
+		}
+
+		// in v01 we actually get spans that we have to transform in traces
```
Removed

djova commented on pkg/trace/api/api.go:
```diff
@@ -263,30 +267,75 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
+		return
+	}
+
+	r.replyTraces(v, w)
```
Good point, updated, but with v04 check so we maintain behavior only sending that for v04.

djova committed:
remove replyTraces

djova committed:
refactor

djova committed:
decoding error test refactor

djova commented on pkg/trace/api/api_test.go:
```diff
@@ -324,6 +324,43 @@ func TestReceiverMsgpackDecoder(t *testing.T) {
 }
 
+func TestReceiverFailedPayloadDecode(t *testing.T) {
+	// tests that decoding errors are correctly tracked in internal statistics
```
Removed

djova commented on pkg/trace/api/api_test.go:
```diff
@@ -324,6 +324,43 @@ func TestReceiverMsgpackDecoder(t *testing.T) {
+		assert.EqualValues(0, r.Stats.GetTagStats(info.Tags{}).TracesDropped.DecodingError)
+	})
+
+	t.Run("DecodingErrorWithTraceCount", func(t *testing.T) {
```
Updated

djova commented on pkg/trace/api/api_test.go:
```diff
@@ -324,6 +324,43 @@ func TestReceiverMsgpackDecoder(t *testing.T) {
+	data := []byte("} invalid json")
+	client := &http.Client{}
+
+	t.Run("DecodingErrorNoTraceCount", func(t *testing.T) {
```
Updated

djova commented on pkg/trace/api/api_test.go:
```diff
@@ -324,6 +324,43 @@ func TestReceiverMsgpackDecoder(t *testing.T) {
+		assert.NoError(err)
+
+		assert.Equal(400, resp.StatusCode)
+		assert.EqualValues(traceCount, r.Stats.GetTagStats(info.Tags{}).TracesDropped.DecodingError)
```
It checks that the decoding error was actually traced. I guess not technically a pure API test, but still seems useful.

djova committed:
cleanup

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+	if ts.Lang != "" {
+		fallbackServiceName = ts.Lang
+	}
+
```
Removed

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+		atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)
+		return fmt.Errorf("SpanID is zero (reason:span_id_zero): %s", s)
+	}
+
```
Removed

djova commented on pkg/trace/api/normalizer.go:
```diff
@@ -21,48 +23,78 @@ const (
+		return fmt.Errorf("SpanID is zero (reason:span_id_zero): %s", s)
+	}
+
 	// Service
```
Removed

djova committed:
reuse WarnString for info message

djova commented on pkg/trace/info/info.go:
```diff
@@ -67,11 +67,11 @@ const (
-    {{if gt $ts.Stats.TracesDropped 0}}
-    WARNING: Traces dropped: {{ $ts.Stats.TracesDropped }}
+    {{ with $ts.Stats.TracesDropped }}
+    WARNING: Traces dropped: {{ .String }}
```
Good point. Updated to reuse `TagStats.WarnString`, which is all in one line already. Then it's also consistent with log summary message.

gbbr replied:
Great! 👍

djova committed:
rename

djova committed:
remove comments

djova committed:
initialize

djova committed:
direct map comparison

djova committed:
direct map comparison

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -248,44 +252,85 @@ func (r *HTTPReceiver) httpHandleWithVersion(v Version, f func(Version, http.Res
-		Interpreter:   req.Header.Get("Datadog-Meta-Lang-Interpreter"),
-		TracerVersion: req.Header.Get("Datadog-Meta-Tracer-Version"),
-	})
+	switch v {
```
This switch case already exists, just reply in there :)

djova committed:
more keys

djova committed:
rename

djova committed:
spacing

djova committed:
maintain OK behavior

djova committed:
refactor handleTraces, it was too big

djova committed:
simplify

djova committed:
fix message

djova committed:
refactor

djova committed:
formatting

djova committed:
simplify

djova committed:
rename

djova committed:
truncate handle utf8

djova committed:
handle large duration

gbbr committed:
Rebase master, simplify API and de-dupe code

gbbr reviewed (dismissed):
I think it's time to merge this. Made a few changes to simplify. Also disabled a test on Windows to not stall work too much, which I'll have to fix as part of another PR (likely https://github.com/DataDog/datadog-agent/pull/3777). PTAL.

gbbr commented:
Flaky test fixed in https://github.com/DataDog/datadog-agent/pull/3790
